### PR TITLE
Request animation frame

### PR DESCRIPTION
**Pull request by coot**
Completes #90.

### CODE REVIEW

File: src/DOM/HTML/Window.js [hunk 1]
```diff
@@ -201,3 +201,33 @@ exports.sessionStorage = function (window) {
...
+  return function(window) {
+    return function() {
+      return window.requestAnimationFrame(fn);
+    };
+  };
+};
+
+exports._cancelAnimationFrame = function(id) {
+  return function(window) {
+    return window.cancelAnimationFrame(id);
```
Comment (garyb) [inline src/DOM/HTML/Window.js]: Need an extra `function() {` wrapping the inner thing here as we're `Eff`-typing these things.

File: src/DOM/HTML/Window.js [hunk 2]
```diff
@@ -201,3 +201,33 @@ exports.sessionStorage = function (window) {
...
+      return function() {
+        return window.requestIdleCallback(fn, opts);
+      };
+    };
+  };
+};
+
+exports._cancelIdleCallback = function(id) {
+  return function(window) {
+    return window.cancelIdleCallback(id);
```
Comment (garyb) [inline src/DOM/HTML/Window.js]: Same with this one!

File: src/DOM/HTML/Window.purs [hunk 3]
```diff
@@ -102,3 +109,31 @@ foreign import scrollY :: forall eff. Window -> Eff (dom :: DOM | eff) Int
 foreign import localStorage :: forall eff. Window -> Eff (dom :: DOM | eff) Storage
 
 foreign import sessionStorage :: forall eff. Window -> Eff (dom :: DOM | eff) Storage
+
+newtype RequestAnimationFrameId = RequestAnimationFrameId Int
+
+derive instance newtypeRequestAnimationFrameId :: Newtype RequestAnimationFrameId _
```
Comment (garyb) [inline src/DOM/HTML/Window.purs]: Can you derive `Eq` and `Ord` for both of the newtypes introduced in here please?

File: src/DOM/HTML/Window.purs [hunk 4]
```diff
@@ -102,3 +109,31 @@ foreign import scrollY :: forall eff. Window -> Eff (dom :: DOM | eff) Int
...
+cancelAnimationFrame :: forall eff. RequestAnimationFrameId -> Window -> Eff (dom :: DOM | eff) Unit
+cancelAnimationFrame idAF = _cancelAnimationFrame (unwrap idAF)
+
+newtype RequestIdleCallbackId = RequestIdleCallbackId Int
+
+derive instance newtypeRequestIdleCallbackId :: Newtype RequestIdleCallbackId _
+
+foreign import _requestIdleCallback :: forall eff. Eff (dom :: DOM | eff) Unit -> { timeout :: Int } -> Window -> Eff (dom :: DOM | eff) Int 
+
+requestIdleCallback :: forall eff. Eff (dom :: DOM | eff) Unit -> { timeout :: Int } -> Window -> Eff (dom :: DOM | eff ) RequestIdleCallbackId
```
Comment (garyb) [inline src/DOM/HTML/Window.purs]: Maybe we should put the options first here? Seems like a more natural ordering for currying perhaps.

Also, is `timeout: 0` different from not specifying a value at all? If so, I guess we'll have to provide two varieties of `requestIdleCallback` to either accept options or not / have `Maybe` properties or something.

**Review by garyb — CHANGES_REQUESTED** [review]
Thanks! Just a couple of things that could use tweaking / discussion.

File: src/DOM/HTML/Window.purs (hunk 4 above)
Comment (coot) [inline src/DOM/HTML/Window.purs]: The spec says that the timeout is only taken into account if it is a [positive integer](https://w3c.github.io/requestidlecallback/#h-the-requestidlecallback-method).

**Comment by garyb** [comment]
Thanks for making those changes, and adding the comment about 0 timeout. Just need those additional updates to the JS (adding the thunk for `Eff`), and also need to swap the arguments in the FFI to reflect the change in the PS file now.

**Comment by coot** [comment]
Ups, I forgot to commit them. Here there are :)

**Review by garyb — APPROVED** [review]

**Comment by garyb** [comment]
Thanks!

**Pull request merged**